storage/shard.go

```diff
@@ -1590,6 +1681,88 @@ func (s *dbShard) removeAnyFlushStatesTooEarly() {
 	s.flushState.Unlock()
 }
 
+func (s *dbShard) SnapshotState() (bool, time.Time) {
+	s.snapshotState.RLock()
+	defer s.snapshotState.RUnlock()
+	return s.snapshotState.isSnapshotting, s.snapshotState.lastSuccessfulSnapshot
+}
+
+func (s *dbShard) markIsSnapshotting() {
+	s.snapshotState.Lock()
+	s.snapshotState.isSnapshotting = true
+	s.snapshotState.Unlock()
+}
+
+func (s *dbShard) markDoneSnapshotting(success bool, completionTime time.Time) {
+	s.snapshotState.Lock()
+	s.snapshotState.isSnapshotting = false
+	if success {
+		s.snapshotState.lastSuccessfulSnapshot = completionTime
+	}
+	s.snapshotState.Unlock()
+}
+
+// CleanupSnapshots examines the snapshot files for the shard that are on disk and
+// determines which can be safely deleted. A snapshot file is safe to delete if it
+// meets one of the following criteria:
+// 		1) It contains data for a block start that is out of retention (as determined
+// 		   by the earliestToRetain argument.)
+// 		2) It contains data for a block start that has already been successfully flushed.
+// 		3) It contains data for a block start that hasn't been flushed yet, but a more
+// 		   recent set of snapshot files (higher index) exists for the same block start.
+// 		   This is because snapshot files are cumulative, so once a new one has been
+//         written out it's safe to delete any previous ones for that block start.
+func (s *dbShard) CleanupSnapshots(earliestToRetain time.Time) error {
```

Comment:
can this moved to the cleanup manager? I know we have the cleanup fileset right below this, we should move that too (subsequent pr maybe?) but might as well not add debt here while we're making changes.

Reply:
I had it like that at one point and it made the cleanup manager code really confusing to follow when it was half-and-half. I have an open issue to move these two methods, so I'd prefer to leave it here and then move them both at once in a separate P.R if possible

Reply:
Sure, that's fair.